.storybook/symbol-doc-loader.js

```diff
@@ -0,0 +1,52 @@
+// @ts-check
+const {createDocProgram} = require('../modules/docs/docgen/createDocProgram');
+const routes = require('./routes');
+
+const routeKeys = Object.keys(routes);
+
+// Tracks files that have been processed. If a file is already processed, it
+// means the file has been updated and we need to update the Typescript program
+// so the changes are reflected in our doc output. If we don't update the TS
+// program, the docs will be processed with the outdated cache of the file
+// contents.
+const filesProcessedMap = new Map();
+
+/** @typedef {import('webpack').loader.Loader} Loader */
+/** @typedef {import('webpack').loader.LoaderContext} LoaderContext */
+
+let {parser, update} = createDocProgram();
+
+/**
+ * @this {LoaderContext}
+ * @param {Parameters<Loader>[0]} source
+ */
+function symbolDocLoader(source) {
+  if (filesProcessedMap.has(this.resourcePath)) {
+    parser = update();
+  }
+  filesProcessedMap.set(this.resourcePath, true);
+
+  const docs = parser.getExportedSymbols(this.resourcePath);
+
+  return (
+    source +
+    `
+  const __docs = ${JSON.stringify(docs, null, '  ')}
+  if (window.__updateDocs) {
+    window.__updateDocs?.(__docs)
+  } else {
+    window.__docs = (window.__docs || []).concat(__docs)
+  }
```

Comment:
These `window` variables are defined in `docs.ts`. The idea is as each file is loaded via Webpack, the global docs array is updated. If a file is updated, `update` is called to inform the Typescript program a source file has changed and it should redo the types. `__updateDocs` will do a replacement of a doc entry if it is already defined. This is how hot reloading updates Storybook.